consensus/src/lib.rs

```diff
@@ -1,27 +1,624 @@
+    fn get_ancestor(&self, block_id: &Id<Block>) -> Result<BlockIndex, BlockError> {
+        let block_index = self.db_tx.get_block_index(block_id).map_err(BlockError::from)?;
+        match block_index {
+            Some(block_index) => {
+                let prev_block =
+                    block_index.get_prev_block_id().as_ref().ok_or(BlockError::NotFound)?;
+                Ok(self.db_tx.get_block_index(prev_block)?.ok_or(BlockError::NotFound)?)
+            }
+            None => Err(BlockError::NotFound),
+        }
+    }
+
+    fn apply_tx_in_undo(&mut self, _tx: &Transaction) {
+        //TODO: not implemented yet
+    }
+
+    fn disconnect_one_block(&mut self, block_index: &mut BlockIndex) -> Result<(), BlockError> {
+        let block = self
+            .db_tx
+            .get_block(
+                //TODO: Remove clone() when Lukas code will be ready
+                block_index.get_block_id().clone(),
+            )?
+            .ok_or(BlockError::Unknown)?;
+        for tx in block.get_transactions().iter().rev() {
+            // TODO: Check that all outputs are available and match the outputs in the block itself
+            //  exactly.
+            for _output in tx.get_outputs() {
+                // if output.is_spendable() {
+                // TODO: We have to check that coin was spent, and that match outputs, height, coinbase
+                // }
+            }
+            // Restore inputs
+            if !tx.is_coinbase() {
+                self.apply_tx_in_undo(tx);
+                //TODO: Check undo status and
+            }
+        }
+        // Not sure what we should return here, might be a new status instead of BlockError?
+        self.db_tx.set_best_block_id(
+            block_index.get_prev_block_id().as_ref().ok_or(BlockError::Unknown)?,
+        )?;
+        // Update connection
+        block_index.next_block_id = None;
+        self.db_tx.set_block_index(block_index)?;
+        Ok(())
+    }
+
+    fn connect_blocks(&mut self, top_block_index: &BlockIndex) -> Result<(), BlockError> {
+        // Disconnect the current chain
+        if let Some(best_block_id) = self.db_tx.get_best_block_id()? {
+            let mut ancestor = self
+                .db_tx
+                .get_block_index(&best_block_id)?
+                .expect("Can't get block index. Inconsistent DB");
+            while !self.is_block_in_main_chain(&ancestor) {
+                self.disconnect_one_block(&mut ancestor)?;
+                ancestor = self.get_ancestor(ancestor.get_block_id())?;
+            }
+        }
+
+        // Connect the new chain
+        let mut ancestor = top_block_index.clone();
+        while !self.is_block_in_main_chain(&ancestor) {
+            self.connect_one_block(&ancestor)?;
+            ancestor = self.get_ancestor(ancestor.get_block_id())?;
+        }
+        Ok(())
+    }
+
+    // Connect mew block
+    fn connect_one_block(&mut self, block: &BlockIndex) -> Result<(), BlockError> {
```

Comment:
This function is called `connect_one_block()` but it takes `&BlockIndex` as parameter and then there's a function called `add_block_index()` which takes `&Block` as parameter. Is there a reason why the functions have been named like this?